internal/sdkv2provider/resource_cloudflare_zone_cache_reserve.go

```diff
@@ -0,0 +1,138 @@
+
+			Requires Cache Reserve subscription.
+		`),
+	}
+}
+
+func resourceCloudflareZoneCacheReserveCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	zoneID := d.Get(consts.ZoneIDSchemaKey).(string)
+
+	// Ensure an unique resource ID to differentiate it from a Zone ID.
+	d.SetId(stringChecksum(fmt.Sprintf("%s/cache-reserve", zoneID)))
+
+	return resourceCloudflareZoneCacheReserveUpdate(ctx, d, meta)
+}
+
+func resourceCloudflareZoneCacheReserveRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get(consts.ZoneIDSchemaKey).(string)
+
+	tflog.Info(ctx, "reading Cache Reserve", map[string]interface{}{
+		"zone_id": zoneID,
+	})
+
+	params := cloudflare.GetCacheReserveParams{}
+	output, err := client.GetCacheReserve(ctx, cloudflare.ZoneIdentifier(zoneID), params)
+	if err != nil {
+		var notFoundError *cloudflare.NotFoundError
+		if errors.As(err, &notFoundError) {
+			tflog.Warn(ctx, "zone could not be found", map[string]interface{}{
+				"zone_id": zoneID,
+			})
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("unable to read Cache Reserve for zone %q: %s", zoneID, err)
+	}
+
+	d.Set("enabled", output.Value == cacheReserveEnabled)
+
+	return nil
+}
+
+func resourceCloudflareZoneCacheReserveUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get(consts.ZoneIDSchemaKey).(string)
+
+	params := cloudflare.UpdateCacheReserveParams{
+		Value: cacheReserveDisabled,
+	}
+	if value, ok := d.GetOk("enabled"); ok && value.(bool) {
+		params.Value = cacheReserveEnabled
+	}
+
+	tflog.Info(ctx, "setting Cache Reserve", map[string]interface{}{
+		"zone_id":       zoneID,
+		"cache_reserve": params.Value == cacheReserveEnabled,
+	})
+
+	_, err := client.UpdateCacheReserve(ctx, cloudflare.ZoneIdentifier(zoneID), params)
+	if err != nil {
+		return diag.Errorf("unable to set Cache Reserve for zone %q: %s", zoneID, err)
+	}
+
+	return resourceCloudflareZoneCacheReserveRead(ctx, d, meta)
+}
+
+func resourceCloudflareZoneCacheReserveDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*cloudflare.API)
+	zoneID := d.Get(consts.ZoneIDSchemaKey).(string)
+
+	tflog.Info(ctx, "deleting Cache Reserve", map[string]interface{}{
+		"zone_id": zoneID,
+	})
+
+	// The Cache Reserve does not have a concept of being added or removed,
+	// it's either turned enabled or disabled, and as such, deleting a Cache
+	// Reserve simply means disabling it, which is also the default.
+	params := cloudflare.UpdateCacheReserveParams{
+		Value: cacheReserveDisabled,
+	}
+	_, err := client.UpdateCacheReserve(ctx, cloudflare.ZoneIdentifier(zoneID), params)
+	if err != nil {
+		return diag.Errorf("unable to delete Cache Reserve for zone %q: %s", zoneID, err)
+	}
+
```

Comment:
we should call `d.SetId("")` after this update too in order to remove it from the state file.

Reply:
On the error path, we probably don't want to call it as it would orphan the resource should we fail to delete it. On the happy path, so to speak, there should be no need to call this explicitly nowadays as it would be called automatically for us when this callback returns without an error.

I would add it in the past, but it seems now this pattern is not desirable per:

- Developer Guide: [Implement delete](https://developer.hashicorp.com/terraform/tutorials/providers/provider-delete)
- https://github.com/hashicorp/terraform-provider-aws/issues/18390
- https://github.com/hashicorp/terraform-provider-aws/pull/4415

Which is why I didn't add it here to the delete callback. Thoughts? Do you still want it added?

Reply:
However, I will add handling the error should the zone no longer be available to make deletion more resilient.

Reply:
> On the happy path, so to speak, there should be no need to call this explicitly nowadays as it would be called automatically for us when this callback returns without an error.

this is a nice DX improvement. thanks for the heads up!